src/reformat.c

```diff
@@ -129,30 +132,35 @@ avifBool avifPrepareReformatState(const avifImage * image, const avifRGBImage *
     return AVIF_TRUE;
 }
 
+// Formulas 20-31 from https://www.itu.int/rec/T-REC-H.273-201612-I/en
 static int avifReformatStateYToUNorm(avifReformatState * state, float v)
 {
-    int unorm = (int)avifRoundf(v * state->yuvMaxChannelF);
+    int unorm = (int)avifRoundf(v * state->yRange + state->yBias);
     unorm = AVIF_CLAMP(unorm, 0, state->yuvMaxChannel);
-    if (state->yuvRange == AVIF_RANGE_LIMITED) {
-        unorm = avifFullToLimitedY(state->yuvDepth, unorm);
-    }
     return unorm;
 }
 
 static int avifReformatStateUVToUNorm(avifReformatState * state, float v)
 {
-    int unorm = (int)avifRoundf(v * state->yuvMaxChannelF);
-    if (state->mode != AVIF_REFORMAT_MODE_IDENTITY) { // Don't adjust U or V when data is raw RGB
-        unorm += state->uvBias;
+    int unorm = 0;
+
+    // For YCgCo, this is only correct for full range
+    if (state->mode == AVIF_REFORMAT_MODE_YCGCO) {
+        assert(state->yuvRange == AVIF_RANGE_FULL);
     }
-    unorm = AVIF_CLAMP(unorm, 0, state->yuvMaxChannel);
-    if (state->yuvRange == AVIF_RANGE_LIMITED) {
-        if (state->mode == AVIF_REFORMAT_MODE_IDENTITY) { // use Y range for all channels when data is raw RGB
-            unorm = avifFullToLimitedY(state->yuvDepth, unorm);
-        } else {
-            unorm = avifFullToLimitedUV(state->yuvDepth, unorm);
```

Comment:
We should also fix avifFullToLimitedY() and avifFullToLimitedUV(). These two functions are implemented as follows:

```
#define FULL_TO_LIMITED(MINLIMITEDY, MAXLIMITEDY, FULLY)                           \
    v = (((v * (MAXLIMITEDY - MINLIMITEDY)) + (FULLY / 2)) / FULLY) + MINLIMITEDY; \
    v = AVIF_CLAMP(v, MINLIMITEDY, MAXLIMITEDY)

int avifFullToLimitedY(int depth, int v)
{
    switch (depth) {
        case 8:
            FULL_TO_LIMITED(16, 235, 255);
            break;
        case 10:
            FULL_TO_LIMITED(64, 940, 1023);
            break;
        case 12:
            FULL_TO_LIMITED(256, 3760, 4095);
            break;
    }
    return v;
}

int avifFullToLimitedUV(int depth, int v)
{
    switch (depth) {
        case 8:
            FULL_TO_LIMITED(16, 240, 255);
            break;
        case 10:
            FULL_TO_LIMITED(64, 960, 1023);
            break;
        case 12:
            FULL_TO_LIMITED(256, 3840, 4095);
            break;
    }
    return v;
}
```

Note that the FULLY arguments are all odd numbers. Some numerical inaccuracy may come from the integer division (FULLY / 2) in the definition of the FULL_TO_LIMITED macro. Could you try this definition?

```
#define FULL_TO_LIMITED(MINLIMITEDY, MAXLIMITEDY, FULLY)                           \
    v = (((v * (MAXLIMITEDY - MINLIMITEDY) * 2) + FULLY) / (FULLY * 2)) + MINLIMITEDY; \
    v = AVIF_CLAMP(v, MINLIMITEDY, MAXLIMITEDY)
```

Reply:
I tested and both gives identical (good enough) result. avifFullToLimitedY() and avifFullToLimitedUV() are doing their best, it's just the result is not suitable as an intermediate value.